src/extension/fileDetectors/contractDetector.ts

```diff
@@ -11,7 +11,7 @@ const LOG_PREFIX = "ContractDetector";
 const SEARCH_PATTERN = "**/*.nef";
 
 type ContractMap = {
-  [contractHash: string]: {
+  [contractName: string]: {
```

Comment:
will this support hash and name?

Reply:
For an internal map, maybe better to use hash but then allow user to specify by name and lookup the hash. Neo express has a custom RPC endpoint ExpressListContracts that you can use to map name -> hash

Reply:
This specific change is just for readability (this map is keyed by names, but the `contractHash` identifier was making the intellisense misleading for me when working on #70)

It would probably be better to use the hash though to avoid collisions  (this can be a separate later work item though).  This data structure is for .nef files found in the workspace--which may not even be deployed yet--so in general I cannot use Neo Express to resolve names to hashes.  I need to find out what algorithm is used to generate the hash from the .nef; is that documented somewhere?

Reply:
> I need to find out what algorithm is used to generate the hash from the .nef; is that documented somewhere?

See: https://github.com/neo-project/neo/blob/master/src/neo/SmartContract/Helper.cs#L84

```

using Neo.Cryptography;
using Neo.Cryptography.ECC;
using Neo.Network.P2P.Payloads;
using Neo.Persistence;
using Neo.SmartContract.Manifest;
using Neo.SmartContract.Native;
using Neo.VM;
using Neo.VM.Types;
using System;
using System.Buffers.Binary;
using System.Collections.Generic;
using System.Linq;

namespace Neo.SmartContract
{
    /// <summary>
    /// A helper class related to smart contract.
    /// </summary>
    public static class Helper
    {
        /// <summary>
        /// The maximum GAS that can be consumed when <see cref="VerifyWitnesses"/> is called.
        /// </summary>
        public const long MaxVerificationGas = 0_50000000;

        /// <summary>
        /// Calculates the verification fee for a signature address.
        /// </summary>
        /// <returns>The calculated cost.</returns>
        public static long SignatureContractCost() =>
            ApplicationEngine.OpCodePrices[OpCode.PUSHDATA1] * 2 +
            ApplicationEngine.OpCodePrices[OpCode.SYSCALL] +
            ApplicationEngine.CheckSigPrice;

        /// <summary>
        /// Calculates the verification fee for a multi-signature address.
        /// </summary>
        /// <param name="m">The minimum number of correct signatures that need to be provided in order for the verification to pass.</param>
        /// <param name="n">The number of public keys in the account.</param>
        /// <returns>The calculated cost.</returns>
        public static long MultiSignatureContractCost(int m, int n)
        {
            long fee = ApplicationEngine.OpCodePrices[OpCode.PUSHDATA1] * (m + n);
            using (ScriptBuilder sb = new())
                fee += ApplicationEngine.OpCodePrices[(OpCode)sb.EmitPush(m).ToArray()[0]];
            using (ScriptBuilder sb = new())
                fee += ApplicationEngine.OpCodePrices[(OpCode)sb.EmitPush(n).ToArray()[0]];
            fee += ApplicationEngine.OpCodePrices[OpCode.SYSCALL];
            fee += ApplicationEngine.CheckSigPrice * n;
            return fee;
        }

        /// <summary>
        /// Check the correctness of the script and ABI.
        /// </summary>
        /// <param name="script">The script of the contract.</param>
        /// <param name="abi">The ABI of the contract.</param>
        public static void Check(byte[] script, ContractAbi abi)
        {
            Check(new Script(script, true), abi);
        }

        /// <summary>
        /// Check the correctness of the script and ABI.
        /// </summary>
        /// <param name="script">The script of the contract.</param>
        /// <param name="abi">The ABI of the contract.</param>
        /// <remarks>Note: The <see cref="Script"/> passed to this method should be constructed with strict mode.</remarks>
        public static void Check(this Script script, ContractAbi abi)
        {
            foreach (ContractMethodDescriptor method in abi.Methods)
                script.GetInstruction(method.Offset);
            abi.GetMethod(string.Empty, 0); // Trigger the construction of ContractAbi.methodDictionary to check the uniqueness of the method names.
            _ = abi.Events.ToDictionary(p => p.Name); // Check the uniqueness of the event names.
        }

        /// <summary>
        /// Computes the hash of a deployed contract.
        /// </summary>
        /// <param name="sender">The sender of the transaction that deployed the contract.</param>
        /// <param name="nefCheckSum">The checksum of the nef file of the contract.</param>
        /// <param name="name">The name of the contract.</param>
        /// <returns>The hash of the contract.</returns>
        public static UInt160 GetContractHash(UInt160 sender, uint nefCheckSum, string name)
        {
            using var sb = new ScriptBuilder();
            sb.Emit(OpCode.ABORT);
            sb.EmitPush(sender);
            sb.EmitPush(nefCheckSum);
            sb.EmitPush(name);

            return sb.ToArray().ToScriptHash();
        }
```

Reply:
This cannot be used for identification of non-deployed contracts though (in a way that will be consistent with the hash after deployment) due to the sender being embedded in the hash.